lib/led-matrix/src/LEDMatrix.cpp

```diff
@@ -0,0 +1,154 @@
+/* LED Matrix Source File */
+
+// naming with m_
+#include "LEDMatrix.h"
+
+#include "mbed.h"
+
+// init thread ptr to nullptr and initialize the requestEndFlashing semaphore to 0 and a max of 1
+LEDMatrix::LEDMatrix(PinName R, PinName G, PinName B)
+    : m_RChannel(R), m_GChannel(G), m_BChannel(B), lightsThread(nullptr), reqEndFlash(0, 1) {}
+
+void LEDMatrix::terminateFlashing() {
+  reqEndFlash.release();  // ask flashing thread to end - signal semaphore
+  lightsThread->join();   // wait for the thread to finish. we've asked it to finish so should be quick wait
+  lightsThread->terminate();
+  delete lightsThread;
+  lightsThread = nullptr;
+}
+void LEDMatrix::flashing() {
+  while (!reqEndFlash.try_acquire()) {  // while we havent been asked to terminate this thread
+    printf("Flashing thread\r\n");
+    // do flashing with flashing_color which is a member var
+
+    while (true) {
+      setColor(flashing_red, flashing_green, flashing_blue);  // is okay that it calls the thread that tries to terminate it... nvmm
+      ThisThread::sleep_for(period_delay);
+      clearLights();
+      ThisThread::sleep_for(period_delay);
+    }
+  }
+}
+
+void LEDMatrix::setColor(uint8_t R, uint8_t G, uint8_t B) {
+  if (lightsThread == nullptr) {  // make sure a thread exists before trying to terminate it
+    terminateFlashing();
+  }
```

Comment:
The logic doesn't follow here. Right now it's saying that if lightsThread is NULL, then terminate flashing?

Reply:
Now that I think about it it, I think that should be `!=`, it could explain why it wasnt doing setColor() properly but at the time I couldn't tell if it was because of the issue with the board itself

Reply:
if you ran this code, you would have gotten a segmentation fault bc of dereferencing a nullptr. im pretty sure the in this case, mbed would cause a msg to be printed. did u see a msg?

Reply:
I dont remember getting a msg. I could be because main ran flashColor() first? So it created a new thread for lightsThread so since lightsThread is never nullptr, it would just never run setColor()... which it didnt